sdk/nodejs/dagger/connect.ts

```diff
@@ -0,0 +1,147 @@
+		stdio: "inherit",
+		cwd: _config.Workdir
+	});
+
+	// Wait for Cloak server to be ready.
+	await waitCloakServer(_config.Port)
+
+	// TODO(TomChv): How transform this part here? This function currently
+	// execute the GQL query but instead we may want to returns a client and
+	// close the server when it's finished.
+	// The only way I see to do this is to let the client hold an optional server
+	// process and shut it down in a Close function if it's defined.
+	// WDYT?
+	//
+	// Piece of code to transform
+	// ```
+	//     await cb(new GraphQLClient(`http://localhost:${this.config.Port}/query`))
+	//       .catch(async (err) => {
+	//         // FIXME:(sipsma) give the engine a sec to flush any progress logs on error
+	//         // Better solution is to send SIGTERM and have a handler in dagger engine that
+	//         // flushes logs before exiting.
+	//         await new Promise((resolve) => setTimeout(resolve, 1000));
+	//         throw err;
+	//       })
+	//       .finally(async () => {
+	//         serverProc.cancel();
+	//         return serverProc.catch((e) => {
+	//           if (!e.isCanceled) {
+	//             console.error("dagger engine error: ", e);
+	//           }
+	//         });
+	//       });
+	// ```
+
+	return new Client();
+}
+
+/**
+ * verifyCloakBinary make sure that user is using the correct version of
+ * Dagger (> 0.3.0).
+ * This function exits on error.
+ */
+async function verifyCloakBinary() {
+	try {
+		execaCommandSync('cloak dev --help');
+	} catch (err) {
+		console.error('⚠️  Please ensure that cloak binary in $PATH is v0.3.0 or newer.');
+		// https://tldp.org/LDP/abs/html/exitcodes.html
+		// | 127 | "command not found" | illegal_command | Possible problem with $PATH or a typo |
+		process.exit(127);
+	}
+}
+
+/**
+ * buildCLIArguments creates a list of string that correspond to
+ * the arguments concatenated to "cloak dev" command.
+ */
+function buildCLIArguments(opts: Required<ConnectOpts>): string[] {
+	const args = [
+		'dev',
+		'--workdir', `${ opts.Workdir }`,
+		'-p', `${ opts.ConfigPath }`,
+		'--port', `${ opts.Port }`,
+	];
+
+	// add local dirs from config in the form of `--local-dir <name>=<path>`
+	for (const [ name, localDir ] of Object.entries(opts.LocalDirs)) {
+		// If path is not absolute, we resolve it to its absolute path
+		// This function do nothing if the path is already absolute
+		const absoluteLocalDirPath = path.resolve(localDir);
+
+		args.push('--local-dir', `${ name }=${ absoluteLocalDirPath }`);
+	}
+
+	return args;
+}
+
+/**
+ * waitCloakServer use an axios client to try connecting to the cloak server
+ * until it successfully connect.
+ * This function has a timeout of 3 minutes to imports and installs all
+ * extensions.
+ */
+async function waitCloakServer(port: number) {
+	const client = axios.create({
+		baseURL: `http://localhost:${port}`,
+	});
+
+	// Time in ms to wait between each attempt.
+	const waitTime = 500;
```

Comment:
nit: having a const here is not necessary IMO. Using the `500` in its place would be more straightforward.

Reply:
Updated in https://github.com/dagger/dagger/pull/3652/commits/8c2c0da71b28e8c37547b10b65cb5fe577583432